configs/example/gem5_library/fdp-hello.py

```diff
@@ -0,0 +1,339 @@
+            cpu = board.get_processor().cores[i].core
+
+            self.l1icaches[i].prefetcher = MultiPrefetcher()
+            if not args.disable_fdp:
+                pf = FetchDirectedPrefetcher(
+                    use_virtual_addresses=True, cpu=cpu
+                )
+                # Optionally register the cache to prefetch into to enable
+                # cache snooping
+                pf.registerCache(self.l1icaches[i])
+                self.l1icaches[i].prefetcher.prefetchers.append(pf)
+
+            self.l1icaches[i].prefetcher.prefetchers.append(
+                TaggedPrefetcher(use_virtual_addresses=True)
+            )
+
+            for pf in self.l1icaches[i].prefetcher.prefetchers:
+                pf.registerMMU(cpu.mmu)
+
+        self.l1dcaches = [
+            L1DCache(size=self._l1d_size)
+            for i in range(board.get_processor().get_num_cores())
+        ]
+        self.l2buses = [
+            L2XBar() for i in range(board.get_processor().get_num_cores())
+        ]
+        self.l2caches = [
+            L2Cache(size=self._l2_size)
+            for i in range(board.get_processor().get_num_cores())
+        ]
+        self.mmucaches = [
+            MMUCache(size="8KiB")
+            for _ in range(board.get_processor().get_num_cores())
+        ]
+
+        self.mmubuses = [
+            L2XBar(width=64)
+            for i in range(board.get_processor().get_num_cores())
+        ]
+
+        if board.has_coherent_io():
+            self._setup_io_cache(board)
+
+        for i, cpu in enumerate(board.get_processor().get_cores()):
+
+            cpu.connect_icache(self.l1icaches[i].cpu_side)
+            self.l1icaches[i].mem_side = self.l2buses[i].cpu_side_ports
+
+            cpu.connect_dcache(self.l1dcaches[i].cpu_side)
+            self.l1dcaches[i].mem_side = self.l2buses[i].cpu_side_ports
+
+            self.mmucaches[i].mem_side = self.l2buses[i].cpu_side_ports
+
+            self.mmubuses[i].mem_side_ports = self.mmucaches[i].cpu_side
+            self.l2buses[i].mem_side_ports = self.l2caches[i].cpu_side
+
+            self.membus.cpu_side_ports = self.l2caches[i].mem_side
+
+            cpu.connect_walker_ports(
+                self.mmubuses[i].cpu_side_ports,
+                self.mmubuses[i].cpu_side_ports,
+            )
+
+            if board.get_processor().get_isa() == ISA.X86:
+                int_req_port = self.membus.mem_side_ports
+                int_resp_port = self.membus.cpu_side_ports
+                cpu.connect_interrupt(int_req_port, int_resp_port)
+            else:
+                cpu.connect_interrupt()
+
+
+cache_hierarchy = CacheHierarchy(
+    l1i_size="32KiB", l1d_size="32KiB", l2_size="1MiB"
+)
+
+
+# 2. Decoupled front-end ------------------------------------------------
+# Next setup the decoupled front-end. Its implemented in the O3 core.
+# Create the processor with one core
+processor = SimpleProcessor(
```

Comment:
As a future PR, we should create a processor & cache hierarchy which has fdp enabled in the stdlib. We'll have to think a little about the interface for this since it requires coordination between the cache hierarchy and the processor.

Reply:
I've now created a draft for this #2595 
The [commit](https://github.com/gem5/gem5/pull/2595/commits/e7b2383948dfaa10e55a917d6998c41986be09f9) adds two components: Processor and CacheHierarchy with decoupling preconfigured.

I think this should be independent of this PR.